File: BOJ/2000-2999/2533.swift
import Foundation

// MARK: - 노드 순서 재정의 - adjacent가 자식인지 부모인지 구분하기 위함
var adjacents = Array(repeating: [Int](), count: 1000001)
var reorderedAdjacents = Array(repeating: [Int](), count: 1000001)
var reorderedIndex = Array(repeating: 0, count: 1000001)
var visit = Array(repeating: false, count: 1000001)
var count = 1

func dfs(index: Int) {
    visit[index] = true
    reorderedIndex[index] = count
    count += 1
    
    for adjacent in adjacents[index] {
        if !visit[adjacent] {
            dfs(index: adjacent)
        }
    }
}

func reorder() {
    dfs(index: 1)       // dfs로 자식 노드 번호가 항상 부모보다 크도록 만듦
    for (index, adj) in adjacents.enumerated() {
        reorderedAdjacents[reorderedIndex[index]] = adj.map{ reorderedIndex[$0] }
    }
}


// MARK: - dp(index, isParentEarly): 부모의 얼리어답터 여부가 isParentEarly일 때, index번 노드를 루트로 하는 부분트리를 완성하는 데 필요한 얼리어답터의 수
let INF = 987654321
var cache = Array(repeating: Array(repeating: INF, count: 1000001), count: 2)

func dp(index: Int, isParentEarly: Bool) -> Int {
    // 기저 사례
    if index != 1 && reorderedAdjacents[index].count == 1 { return isParentEarly ? 0 : 1 }  // 리프 노드인 경우
    
    // 캐싱
    var ret = cache[isParentEarly ? 1 : 0][index]
    if ret != INF { return ret }
    
    var early = 1
    var noEarly = 0
    for child in reorderedAdjacents[index] {
        guard child > index else { continue }
        
        early += dp(index: child, isParentEarly: true)
        noEarly += dp(index: child, isParentEarly: false)
    }
    
    if isParentEarly {              // 부모가 얼리어답터인 경우 - 얼리어답터이지 않아도 됨
        ret = min(early, noEarly)
    } else {                        // 부모가 얼리어답터가 아닌 경우 - 무조건 얼리어답터여야만 함
        ret = early
    }
    
    cache[isParentEarly ? 1 : 0][index] = ret
    
    return ret
}


let N = Int(readLine()!)!
for _ in (0..<N-1) {
    let input = readLine()!.split(separator: " ").map{ Int(String($0))! }
    let A = input[0]
    let B = input[1]
    
    adjacents[A].append(B)
    adjacents[B].append(A)
}

reorder()

print(dp(index: 1, isParentEarly: true))

File: BOJ/2000-2999/2533.cpp
#include <iostream>
#include <cstring>
#include <vector>

using namespace std;

int N;
vector<int> adjacents[1000001];
vector<int> reorderedAdjacents[1000001];
int reorderedIndex[1000001];
bool visit[1000001];
int cnt = 1;

void dfs(int index) {
    visit[index] = true;
    reorderedIndex[index] = cnt++;
    
    for(int i = 0; i < adjacents[index].size(); i++) {
        int adjacent = adjacents[index][i];
        if (!visit[adjacent]) {
            dfs(adjacent);
        }
    }
}

void reorder() {
    dfs(1);
    
    for(int i=1; i <= N; i++) {
        for(int j=0; j < adjacents[i].size(); j++) {
            reorderedAdjacents[reorderedIndex[i]].push_back(reorderedIndex[adjacents[i][j]]);
        }
    } 
}

int INF = 987654321;
int cache[2][1000001];

int dp(int idx, int isParentEarly) {
	if (idx != 1 && reorderedAdjacents[idx].size() == 1) {
	    return !isParentEarly;
	}

	int& ret = cache[isParentEarly][idx];
	if (ret != INF) return ret;

	int early = 1;
	int noEarly = 0;
	for(int i = 0; i < reorderedAdjacents[idx].size(); i++) {
	    int adjacent = reorderedAdjacents[idx][i];
	    if (adjacent > idx) {
	        early += dp(adjacent, true);
	        noEarly += dp(adjacent, false);
	    }
	}
	
	if (isParentEarly) {
	    ret = min(early, noEarly);
	} else {
	    ret = early;
	}

	return ret;
}

int main() {
	ios::sync_with_stdio(false);
	cin.tie(NULL); cout.tie(NULL);

	for (int i = 0; i < 1000001; i++) {
	    for (int j = 0; j < 2; j++) {
	        cache[j][i] = INF;
	    }
	}

	cin >> N;
	for (int i = 0; i < N - 1; i++) {
		int A, B;
		cin >> A >> B;
		adjacents[A].push_back(B);
		adjacents[B].push_back(A);
	}
	
	reorder();

	cout << dp(1, 1);
}
